### ISSUE TITLE
[rllib] SAC/DDPG state preprocessor variables not included in grad calculation

### ISSUE THREAD

**Issue by bentzinir** [ray-project/ray#10765]
### What is the problem?
It is not clear to me if the variables of a state-preprocessor, i.e. the core "forward" layers, are assigned to a gradient.
Assume I use sac (or ddpg) with a VisionNet as a state-preprocessor. 
Those variables are not registered as actor variables, nor as q variables.
However, the actor (critic) gradient is selectively defined over actor (critic) variables only.
Is it possible that the variables of the state-preprocessor are not being trained?

*Ray version and other system information (Python version, TensorFlow version, OS):*
python3.6, ray 0.8, ubuntu 18

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by ericl** [comment]
Shouldn't they be registered by default by being members of nn.Module? You can probably check the .variables() list to see if they are indeed registered.

**Comment by bentzinir** [comment]
Forgive the misleading title. They are registered but do not belong to any gradient.
The sac implementation defines its gradients over a selected list of variables.
For example, the actor gradient is defined explicitly over the actor variables (see `gradients_fn` in sac_tf_policy.py`). Those are the weights that are returned by `policy_variables` in `sac_tf_model.py`. However, this contains just the weights of the output heads and not the weights of the state-preprocessor.
While there are 24 registered variables, most of them belong to the cnn, the gradient of the policy is defined over 4 weights only.

**Comment by ericl** [comment]
Good point. That does look like a problem, @michaelzhiluo can you have a closer look?

**Comment by michaelzhiluo** [comment]
It makes sense that the preprocessor variables aren't part of the model, as intended in [data pipeline](https://docs.ray.io/en/latest/rllib-models.html). Is it possible you can move this preprocessor the model (i.e. define a custom model)?

**Comment by bentzinir** [comment]
The current data pipeline supports a shared body configuration because the state-preprocessor is assumed to be shared between the actor and the critic. I think this should be left as a design choice.

**Comment by ericl** [comment]
@michaelzhiluo , that pipeline is referring to the preprocessor outside the model. @bentzinir is talking about the state_preprocessor network in SAC which is differentiable through the loss.

**Comment by sven1977** [comment]
Yes, this is a bug. The "preprocessor" here is not an RLlib Preprocessor object, but e.g. a simple Conv2D stack (chosen automatically if possible by RLlib) pre-attached to the Q- and policy models.
Will fix this. ...

**Comment by sven1977** [comment]
Sorry, can you be more specific about the version of ray you are using, as well as some information on your config, tf/pytorch, etc..?

**Comment by sven1977** [comment]
@bentzinir

**Comment by sven1977** [comment]
I do see all variables registered correctly for tf as well as all parameters included in the nn.Module for pytorch:
`rllib\agents\sac\tests\test_sac.py::TestSAC::test_sac_compilation`
framework=torch|tf
env=MsPacmanNoFrameskip-v4

**Comment by ericl** [comment]
@sven1977 see the comment https://github.com/ray-project/ray/issues/10765#issuecomment-696537955, it seems the vars are registered but not included in the grad calcuation?

**Comment by bentzinir** [comment]
ray 0.8, TensorFlow 1. I think that this is also the case with ddpg.

**Comment by sven1977** [comment]
Ah, yes, sorry, checking the custom gradient functions ...

**Comment by sven1977** [comment]
Yeah, we don't calculate or apply gradients for these, true. Ok, I'll make the `self._actor_optimizer` take care of training the preprocessor vars. Any objections?

**Comment by sven1977** [comment]
https://github.com/ray-project/ray/pull/10959

**Comment by sven1977** [comment]
DDPG still missing. Will PR (see above) tomorrow.

**Comment by bentzinir** [comment]
> Yeah, we don't calculate or apply gradients for these, true. Ok, I'll make the `self._actor_optimizer` take care of training the preprocessor vars. Any objections?

Makes sense. But the main issue IMOP is to allow a separate body configuration.
Another thing, you might want to take the minimum of the Qs in the discrete actor loss function.
Nir

**Comment by sven1977** [comment]
So true. We are planning on making Model-architecture configurations much simpler and more flexible in the future (~Q4, Q1 2021). What you should do if you want to separate into 2 Conv2D stacks is to just write a custom Model class that does this, then wrap the SACModel with that custom class.

**Comment by sven1977** [comment]
Great catch on the min Q for discrete actions! I'll fix that.

**Comment by FabianSchuetze** [comment]
I also had problems with this and assigned the variables to the critic loss, not the actor loss. My motivation for this was based on empirical evidence from the [deepmind control suite paper](https://arxiv.org/pdf/1801.00690.pdf). Figure 7 and 8 show that the network barely learns anything when the CNN variables are optimized as part of the actor loss.

**Comment by bentzinir** [comment]
That's interesting. They say that separate bodies performed the worse, which is what I use now :(
In my opinion, though, such conclusions should be treated with caution.
I couldn't find Figure 8. I assume you referred to Fig7?

Any objections to closing the issue?

**Comment by FabianSchuetze** [comment]
@bentzinir  : No objection to closing - thanks for your comment! Yes, sorry, I meant Figure 6 and 7.

Did you have any success with training a pixel based policy other than on Atari? So far, its eluding me and I would be grateful for any hints which environments and algorithms work!

**Comment by bentzinir** [comment]
It would be foolish of me to pretend to know the answer to this...
But here is my two cents:
Pixel-based policies usually go together with discrete actions. In this case, PPO is your go-to algorithm.
As a trust-region algorithm, it will provide you with "safe" optimization.
However, as an on-policy algorithm, you might be facing exploration challenges that are better to solve by tweaking the environment if possible.
